language/move-stdlib/src/natives/mod.rs

```diff
@@ -7,90 +7,164 @@ pub mod event;
 pub mod hash;
 pub mod signer;
 pub mod string;
-pub mod vector;
-
 #[cfg(feature = "testing")]
 pub mod unit_test;
+pub mod vector;
 
 #[cfg(feature = "testing")]
 pub mod debug;
 
+mod helpers;
+
 use move_core_types::account_address::AccountAddress;
-use move_vm_runtime::{
-    native_functions,
-    native_functions::{NativeFunction, NativeFunctionTable},
-};
-
-pub fn all_natives(move_std_addr: AccountAddress) -> NativeFunctionTable {
-    const NATIVES: &[(&str, &str, NativeFunction)] = &[
-        ("bcs", "to_bytes", bcs::native_to_bytes),
-        ("event", "write_to_event_store", event::write_to_event_store),
-        ("hash", "sha2_256", hash::native_sha2_256),
-        ("hash", "sha3_256", hash::native_sha3_256),
-        ("signer", "borrow_address", signer::native_borrow_address),
-        ("string", "internal_check_utf8", string::native_check_utf8),
-        (
-            "string",
-            "internal_is_char_boundary",
-            string::native_is_char_boundary,
-        ),
-        ("string", "internal_sub_string", string::native_sub_string),
-        ("string", "internal_index_of", string::native_index_of),
-        ("vector", "length", vector::native_length),
-        ("vector", "empty", vector::native_empty),
-        ("vector", "borrow", vector::native_borrow),
-        ("vector", "borrow_mut", vector::native_borrow),
-        ("vector", "push_back", vector::native_push_back),
-        ("vector", "pop_back", vector::native_pop),
-        ("vector", "destroy_empty", vector::native_destroy_empty),
-        ("vector", "swap", vector::native_swap),
-        #[cfg(feature = "testing")]
-        ("debug", "print", debug::native_print),
-        #[cfg(feature = "testing")]
-        (
-            "debug",
-            "print_stack_trace",
-            debug::native_print_stack_trace,
-        ),
-        #[cfg(feature = "testing")]
-        (
-            "unit_test",
-            "create_signers_for_testing",
-            unit_test::native_create_signers_for_testing,
-        ),
-    ];
-    native_functions::make_table(move_std_addr, NATIVES)
+use move_vm_runtime::native_functions::{make_table_from_iter, NativeFunctionTable};
+
+#[derive(Debug, Clone)]
+pub struct GasParameters {
+    bcs: bcs::GasParameters,
+    hash: hash::GasParameters,
+    signer: signer::GasParameters,
+    string: string::GasParameters,
+    vector: vector::GasParameters,
+
+    #[cfg(feature = "testing")]
+    unit_test: unit_test::GasParameters,
+}
+
+impl GasParameters {
+    pub fn zeros() -> Self {
```

Comment:
What is this used for in most cases? Can we copy over the initial schedule from whatever was there before? Or is there some reason not to?

Reply:
Good question. 

> Can we copy over the initial schedule from whatever was there before?
I thought about doing this previously and actually listed this as a work item. However after inspecting the values, it felt to me that some of the values were so absurd and inconsistent that I was no longer sure if I should even bother keeping this old schedule. Here are some examples:
1. `vector::swap` costs 1437 gas units -- for simply doing a shallow swap? `vector::borrow` costs 1335 -- for creating a simple reference? In comparison, `vector::empty` costs 85; `vector::length` costs 99; `vector::destroy_empty` costs 573.
  a. It's weird that operations that allocate memories cost even less than some of the ones that only touches a few words.
2. `vector::push_back` costs 54 * size_of(val being pushed) while `vector::pop_back` costs a constant 228 units
3. The string natives all costs some constant numbers even though most operations are linear

I took a look at the callsites of `GasParameters::zeros` and it seemed to me they can be classified into 3 types:
1. Tests that don't care about gas metering -- using the all-zero gas schedule is the right thing to do
2. Tests that may care about gas metering but don't yet -- we can come up with something that actually makes sense to them when we start to care
3. Tools (move-cli/unit tests) -- these really shouldn't have a hardcoded gas schedule. I have left some TODOs to remind us that we'll need to redesign these parts in the future.